client.js

```diff
@@ -102,67 +82,98 @@ async function insert({ _, ...rest }) {
-    `${lastNode.ip}:${lastNode.port}`,
-    grpc.credentials.createInsecure()
-  );
-  // The argument is total garbage
-  client.getSuccessor_remotehelper({ id: 99 }, (err, node) => {
-    if (err) {
-      console.log("Error is : ", err);
-      let nodeToDelete = lastNode.id;
-      // Remove the node from the bucket and select a random node
-      for (elem in Object.keys(bigBucketOfData)) {
-        if (
-          elem &&
-          bigBucketOfData[elem] &&
-          bigBucketOfData[elem].successor &&
-          bigBucketOfData[elem].successor.id &&
-          bigBucketOfData[elem].successor.id !== lastNode.id
-        ) {
-          lastNode = bigBucketOfData[elem].successor;
-          break;
-        }
-      }
-      delete bigBucketOfData[nodeToDelete];
-    } else {
-      if (bigBucketOfData[lastNode.id]) {
-        bigBucketOfData[lastNode.id].successor = node;
+class ChordCrawler {
+  constructor(ip, port, stepInSeconds) {
+    this.ip = ip;
+    this.port = port;
+    this.state = {};
+    this.ring = new Set([]);
+    setInterval(async () => {
+      await this.crawl();
+    }, stepInSeconds);
+  }
+  static buildConnectionString(ip, port) {
+    return `${ip}:${port}`;
+  }
+  getConnectionString() {
+    return ChordCrawler.buildConnectionString(this.ip, this.port);
+  }
+  async crawl() {
+    const connectionString = this.getConnectionString();
+    console.log(`Connecting to ${connectionString}`);
+    const client = caller(connectionString, PROTO_PATH, "Node");
+
+    try {
+      const DUMMY_REQUEST_OBJECT = { id: 99 };
+      const successorNode = await client.getSuccessor_remotehelper(
+        DUMMY_REQUEST_OBJECT
+      );
+      // If we have an entry for the current node we're connecting to, update the successor
+      // If not, this is likley the initial gRPC call, so just skip. We'll get this after our first complete walk
+      if (this.state[connectionString]) {
+        this.state[connectionString].successor = successorNode;
       }
 
-      // If we've walked the logical ring and we didn't touch a node, delete it
-      if (ring.has(node.id)) {
-        for (elem of Object.keys(bigBucketOfData)) {
-          console.log(bigBucketOfData[elem]);
-          if (!ring.has(bigBucketOfData[elem].id)) {
-            delete bigBucketOfData[elem];
+      // If we touch a node we've already touched, we've walked a complete cycle around the overlay network
+      // We can now flush dangling nodes not encountered during the walk
+      if (this.ring.has(connectionString)) {
+        for (let storedConnectionString of Object.keys(this.state)) {
+          if (!this.ring.has(storedConnectionString)) {
+            delete this.state[storedConnectionString];
           }
         }
-        ring.clear();
+        // Clear because we are starting a new walk around the overlay network
+        this.ring.clear();
+      } else {
+        // Add the node to the walk
+        this.ring.add(connectionString);
       }
```

Comment:
Maybe put this into a function with a meaningful name instead of having that long comment?